File: mobile/lib/network/discovery.dart
import 'package:multicast_dns/multicast_dns.dart';

import 'wheeldeck_client.dart';

/// The DNS-SD service type the desktop advertises under.
///
/// Discovery (TASK-030) queries this name; the desktop's advertiser will
/// register the same string so both sides stay in sync.
const String wheelDeckServiceType = '_wheeldeck._tcp.local';

/// A desktop server resolved through mDNS or entered manually.
class DiscoveredServer {
  const DiscoveredServer({
    required this.host,
    required this.port,
    required this.name,
  });

  /// IP address the WebSocket listener is reachable at.
  final String host;

  /// WebSocket port the desktop server listens on.
  final int port;

  /// Human-readable service instance name, for the selection UI.
  final String name;

  ConnectionTarget toConnectionTarget() => ConnectionTarget(
        mode: ConnectionMode.autoDiscover,
        ipAddress: host,
        port: port,
      );
}

/// Resolves advertised WheelDeck servers to dialable targets.
typedef ServerResolver = Future<List<DiscoveredServer>> Function();

/// Discovers desktop servers via mDNS and exposes manual-IP entry as the
/// fallback for networks that block multicast (public Wi-Fi with client
/// isolation, per the PRD).
class ServerDiscovery {
  ServerDiscovery({ServerResolver? resolve})
      : _resolve = resolve ?? MulticastResolver().resolve;

  final ServerResolver _resolve;

  /// Queries the network for advertised WheelDeck servers.
  Future<List<DiscoveredServer>> discover() => _resolve();

  /// Builds a manual target from a user-entered host and optional port.
  static ConnectionTarget manualTarget({
    required String host,
    int? port,
  }) =>
      ConnectionTarget(
        mode: ConnectionMode.manual,
        ipAddress: host,
        port: port ?? WheelDeckClient.defaultPort,
      );
}

/// Drives the [MDnsClient] through the PTR -> SRV -> A resolution chain.
class MulticastResolver {
  Future<List<DiscoveredServer>> resolve() async {
    final client = MDnsClient();
    try {
      await client.start();

      final pointers = await client
          .lookup<PtrResourceRecord>(
            ResourceRecordQuery.serverPointer(wheelDeckServiceType),
          )
          .toList();

      final servers = <DiscoveredServer>[];
      for (final pointer in pointers) {
        final server = await _resolveInstance(client, pointer.domainName);
        if (server != null) {
          servers.add(server);
        }
      }

      return servers;
    } finally {
      client.stop();
    }
  }

  Future<DiscoveredServer?> _resolveInstance(
    MDnsClient client,
    String instanceName,
  ) async {
    final services = await client
        .lookup<SrvResourceRecord>(
          ResourceRecordQuery.service(instanceName),
        )
        .toList();

    if (services.isEmpty) {
      return null;
    }

    final service = services.first;
    final addresses = await client
        .lookup<IPAddressResourceRecord>(
          ResourceRecordQuery.addressIPv4(service.target),
        )
        .toList();

    if (addresses.isEmpty) {
      return null;
    }

    return DiscoveredServer(
      host: addresses.first.address.address,
      port: service.port,
      name: _displayName(instanceName),
    );
  }

  String _displayName(String instanceName) {
    final suffix = '.$wheelDeckServiceType';
    return instanceName.endsWith(suffix)
        ? instanceName.substring(0, instanceName.length - suffix.length)
        : instanceName;
  }
}

File: android/app/src/main/kotlin/dev/fazrigading/wheeldeck/data/services/Discovery.kt
package dev.fazrigading.wheeldeck.data.services

import android.content.Context
import android.net.nsd.NsdManager
import android.net.nsd.NsdServiceInfo
import dev.fazrigading.wheeldeck.domain.models.ConnectionMode
import dev.fazrigading.wheeldeck.domain.models.ConnectionTarget
import dev.fazrigading.wheeldeck.domain.models.DefaultWheelDeckPort
import dev.fazrigading.wheeldeck.domain.models.DiscoveredServer
import java.util.concurrent.CopyOnWriteArrayList
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay

/// The DNS-SD service type the desktop advertises under.
///
/// Android's NsdManager wants the type without the `.local` domain but with a
/// trailing dot; the Dart app used `_wheeldeck._tcp.local`. This constant is
/// the single place where the Android format lives.
const val WheelDeckServiceType = "_wheeldeck._tcp."

/// Resolves advertised WheelDeck servers to dialable targets.
fun interface ServerResolver {
    suspend fun resolve(): List<DiscoveredServer>
}

/// Discovers desktop servers via mDNS and exposes manual-IP entry as the
/// fallback for networks that block multicast (public Wi-Fi with client
/// isolation, per the PRD).
class ServerDiscovery(private val resolver: ServerResolver) {
    constructor(context: Context) : this(NsdResolver(context))

    /// Queries the network for advertised WheelDeck servers.
    suspend fun discover(): List<DiscoveredServer> = resolver.resolve()

    companion object {
        /// Builds a manual target from a user-entered host and optional port.
        fun manualTarget(host: String, port: Int? = null) = ConnectionTarget(
            mode = ConnectionMode.Manual,
            ipAddress = host,
            port = port ?: DefaultWheelDeckPort,
        )
    }
}

/// Drives [NsdManager] through discovery and per-instance resolution.
/// ponytail: fixed 3s one-shot sweep window; switch to continuous discovery
/// with onServiceLost updates if stale entries become a real problem.
class NsdResolver(context: Context, private val discoveryWindowMillis: Long = 3_000) : ServerResolver {
    private val nsManager = context.getSystemService(Context.NSD_SERVICE) as NsdManager

    override suspend fun resolve(): List<DiscoveredServer> = coroutineScope {
        val servers = CopyOnWriteArrayList<DiscoveredServer>()
        val started = CompletableDeferred<Unit>()
        val listener = object : NsdManager.DiscoveryListener {
            override fun onDiscoveryStarted(serviceType: String) {
                started.complete(Unit)
            }

            override fun onStartDiscoveryFailed(serviceType: String, errorCode: Int) {
                started.completeExceptionally(
                    IllegalStateException("NSD discovery failed to start: $errorCode"),
                )
            }

            override fun onServiceFound(serviceInfo: NsdServiceInfo) {
                nsManager.resolveService(
                    serviceInfo,
                    object : NsdManager.ResolveListener {
                        override fun onResolveFailed(info: NsdServiceInfo, errorCode: Int) {}

                        override fun onServiceResolved(info: NsdServiceInfo) {
                            servers.add(info.toDiscoveredServer())
                        }
                    },
                )
            }

            override fun onServiceLost(serviceInfo: NsdServiceInfo) {}

            override fun onDiscoveryStopped(serviceType: String) {}

            override fun onStopDiscoveryFailed(serviceType: String, errorCode: Int) {}
        }

        nsManager.discoverServices(WheelDeckServiceType, NsdManager.PROTOCOL_DNS_SD, listener)
        try {
            started.await()
            delay(discoveryWindowMillis)
        } finally {
            runCatching { nsManager.stopServiceDiscovery(listener) }
        }
        servers.toList()
    }

    private fun NsdServiceInfo.toDiscoveredServer() = DiscoveredServer(
        host = host?.hostAddress.orEmpty(),
        port = port,
        name = serviceName,
    )
}
